build.gradle

```diff
@@ -17,54 +17,61 @@
+        all {
+          pom.withXml { XmlProvider xml ->
+            Node node = xml.asNode()
+            node.appendNode('inceptionYear', '2009')
 
-        licenses {
-          license {
-            name 'The Apache Software License, Version 2.0'
-            url 'http://www.apache.org/licenses/LICENSE-2.0.txt'
-            distribution 'repo'
+            Node license = node.appendNode('licenses').appendNode('license')
+            license.appendNode('name', 'The Apache Software License, Version 2.0')
+            license.appendNode('url', 'http://www.apache.org/licenses/LICENSE-2.0.txt')
+            license.appendNode('distribution', 'repo')
           }
         }
       }
-    }
-    extraArchive {
-      javadoc = true
-      tests = false
-    }
-    // we have our own username/password prompts so that they only happen once
-    // TODO: add gpg signing prompts
-    project.gradle.taskGraph.whenReady { taskGraph ->
-      if (taskGraph.allTasks.any { it.name == 'uploadArchives' }) {
-        Console console = System.console()
-        if (project.hasProperty('nexusUsername') == false) {
-          String nexusUsername = console.readLine('\nNexus username: ')
-          project.rootProject.allprojects.each {
-            it.ext.nexusUsername = nexusUsername
-          }
+      repositories.maven {
+        name 'sonatype'
+        if (version.endsWith('-SNAPSHOT')) {
+          url 'https://oss.sonatype.org/content/repositories/snapshots/'
+        } else {
+          url 'https://oss.sonatype.org/service/local/staging/deploy/maven2/'
         }
-        if (project.hasProperty('nexusPassword') == false) {
-          String nexusPassword = new String(console.readPassword('\nNexus password: '))
-          project.rootProject.allprojects.each {
-            it.ext.nexusPassword = nexusPassword
+
+        // It would be nice to pass a custom impl of PasswordCredentials
+        // that could lazily read username/password from the console if not
+        // present as properties. However, gradle's credential handling is
+        // completely broken for custom impls. It checks that the class
+        // passed in is exactly PasswordCredentials or AwsCredentials.
+        // So instead, we must rely on heuristics of "are we publishing"
+        // by inspecting the command line, stash the credentials
+        // once read in the root project, and set them on each project
+        if (gradle.startParameter.taskNames.contains('publish')) {
+          Console console = System.console()
+          if (project.rootProject.hasProperty('nexusUsername') == false) {
+            project.rootProject.ext.nexusUsername = console.readLine('\nNexus username: ')
+          }
+          if (project.rootProject.hasProperty('nexusPassword') == false) {
+            project.rootProject.ext.nexusPassword = new String(console.readPassword("\nNexus password: "))
+          }
+
+          credentials {
+            username = project.rootProject.nexusUsername 
```

Comment:
These are already being read as gradle properties (both now and before). You can either add them on the command line with eg `gradle -PnexusUsername=myuser publish`, or you can add `nexusUsername=myuser` in your home gradle properties eg `~/.gradle/gradle.properties`.